qubesmanager/qube_manager.py

```diff
@@ -61,140 +71,397 @@ def mousePressEvent(self, e):  # pylint: disable=invalid-name
+            if col_name == "State":
+                return vm.state
+            if col_name == "Template":
+                if vm.template is None:
+                    return vm.klass
+                return vm.template
+            if col_name == "NetVM":
+                return vm.netvm
+            if col_name == "Disk Usage":
+                return vm.disk
+            if col_name == "Internal":
+                return "Yes" if vm.internal else ""
+            if col_name == "IP":
+                return vm.ip
+            if col_name == "Include in backups":
+                return "Yes" if vm.inc_backup else ""
+            if col_name == "Last backup":
+                return vm.last_backup
+            if col_name == "Default DispVM":
+                return vm.dvm
+            if col_name == "Is DVM Template":
+                return "Yes" if vm.dvm_template else ""
+            if col_name == "Virt Mode":
+                return vm.virt_mode
+        if role == Qt.DecorationRole:
+            if col_name == "Type":
+                try:
+                    return self.klass_pixmap[vm.klass]
+                except KeyError:
+                    pixmap = QPixmap()
+                    icon_name = ":/"+vm.klass.lower()+".png"
+                    icon_name = icon_name.replace("adminvm", "dom0")
+                    icon_name = icon_name.replace("dispvm", "appvm")
+                    pixmap.load(icon_name)
+                    self.klass_pixmap[vm.klass] = pixmap.scaled(icon_size)
+                    return self.klass_pixmap[vm.klass]
+
+            if col_name == "Label":
+                try:
+                    return self.label_pixmap[vm.label]
+                except KeyError:
+                    icon = QIcon.fromTheme(vm.label.icon)
+                    self.label_pixmap[vm.label] = icon.pixmap(icon_size)
+                    return self.label_pixmap[vm.label]
+
+        if role == Qt.FontRole:
+            if col_name == "Template":
+                if vm.template is None:
+                    font = QFont()
+                    font.setItalic(True)
+                    return font
+
+        if role == Qt.ForegroundRole:
+            if col_name == "Template":
+                if vm.template is None:
+                    return QColor("gray")
+
+        # Used for get VM Object
+        if role == Qt.UserRole:
+            return vm
+
+        # Used for sorting
+        ret = None
+        if role == Qt.UserRole + 1:
+            if vm.qid == 0:
+                ret = ""
+            elif col_name == "Type":
+                ret = vm.klass
+            elif col_name == "Label":
+                ret = vm.label.name
+            elif col_name == "State":
+                ret = str(vm.state)
+            elif col_name == "Disk Usage":
+                ret = vm.disk_float
+            else:
+                ret = self.data(index, Qt.DisplayRole)
+
+            if col_name != "Name":
+                ret = ret + vm.name
```

Comment:
The tests checks if dom0 is sorted first, but this doesn't guarantee it. In fact, many of those tests fails for me. For example when sorting by "Template", it gets `dom0` vs `debian-10-xfcesys-gui` and the latter is sorted earlier.

Reply:
Oh, it was ' return "" ' for dom0 and worked fine but I changed it fixing a pylint warning :rofl:

Reply:
This concatenation has other unintended effects:
```
======================================================================
FAIL: test_100_sorting (qubesmanager.tests.test_qube_manager.QubeManagerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/qubesmanager/tests/test_qube_manager.py", line 251, in test_100_sorting
    self.__check_sorting("Template")
  File "/usr/lib/python3.8/site-packages/qubesmanager/tests/test_qube_manager.py", line 1266, in __check_sorting
    self.assertGreater(
AssertionError: 'fedora-32' not greater than 'fedora-32-xfce' : Incorrect sorting for Template

----------------------------------------------------------------------
```

That's because it isn't `fedora-32` sorted against `fedora-32-xfce` - it is `fedora-32vmname` sorted against `fedora-32-xfcevmname` and indeed this reverses the order here. Adding some separator that cannot appear in the column content (`!`? a space?) should help, but still feels hacky. Perhaps use tuples (primary value, vmname) to make sure the vm name is used for sorting really only when the first value matches?

Reply:
I tried with a better way, or a least not so hacky.